File: evernote_client/mcp/read_tools.py
"""Read-only MCP tools for Evernote."""

from __future__ import annotations

from evernote_client import service
from evernote_client.models import (
    NotebookInfo,
    NoteContent,
    NoteMetadata,
    SearchResult,
    TagInfo,
)
from evernote_client.service import PrivateNoteError

from .app import mcp


@mcp.tool()
def search_notes(
    query: str = "",
    notebook_name: str = "",
    tags: list[str] | None = None,
    max_results: int = 20,
    offset: int = 0,
) -> SearchResult:
    """Search notes using Evernote search grammar.

    Args:
        query: Search query (Evernote search grammar)
        notebook_name: Filter by notebook name
        tags: Filter by tag names
        max_results: Maximum number of results (default 20, max 100)
        offset: Offset for pagination
    """
    return service.search_notes(
        query=query,
        notebook_name=notebook_name,
        tags=tags,
        max_results=max_results,
        offset=offset,
    )


@mcp.tool()
def get_note(guid: str) -> NoteMetadata | str:
    """Get note metadata (title, tags, notebook, dates).

    Args:
        guid: Note GUID
    """
    try:
        return service.get_note(guid)
    except PrivateNoteError:
        return "Access denied: note is private."


@mcp.tool()
def get_note_content(guid: str) -> NoteContent | str:
    """Get full note content as Markdown.

    Args:
        guid: Note GUID
    """
    try:
        return service.get_note_content(guid)
    except PrivateNoteError:
        return "Access denied: note is private."


@mcp.tool()
def list_notebooks() -> list[NotebookInfo]:
    """List all notebooks with guid, name, and stack."""
    return service.list_notebooks()


@mcp.tool()
def list_tags() -> list[TagInfo]:
    """List all tags with guid and name."""
    return service.list_tags()

File: src/server/read-tools.ts
import type { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { z } from 'zod/v4';
import * as service from '../service.js';

export function registerReadTools(server: McpServer): void {
  server.tool(
    'search_notes',
    'Search notes using Evernote search grammar',
    {
      query: z.string().optional().describe('Search query (Evernote search grammar)'),
      notebook_name: z.string().optional().describe('Filter by notebook name'),
      tags: z.array(z.string()).optional().describe('Filter by tag names'),
      max_results: z.number().optional().describe('Maximum number of results (default 20, max 100)'),
      offset: z.number().optional().describe('Offset for pagination'),
    },
    async (args) => {
      const result = await service.searchNotes(
        args.query ?? '',
        args.notebook_name ?? '',
        args.tags ?? null,
        args.max_results ?? 20,
        args.offset ?? 0
      );
      return { content: [{ type: 'text', text: JSON.stringify(result, null, 2) }] };
    }
  );

  server.tool(
    'get_note',
    'Get note metadata (title, tags, notebook, dates)',
    {
      guid: z.string().describe('Note GUID'),
    },
    async (args) => {
      const result = await service.getNote(args.guid);
      return { content: [{ type: 'text', text: JSON.stringify(result, null, 2) }] };
    }
  );

  server.tool(
    'get_note_content',
    'Get full note content as Markdown',
    {
      guid: z.string().describe('Note GUID'),
    },
    async (args) => {
      const result = await service.getNoteContent(args.guid);
      return { content: [{ type: 'text', text: JSON.stringify(result, null, 2) }] };
    }
  );

  server.tool(
    'list_notebooks',
    'List all notebooks with guid, name, and stack',
    {},
    async () => {
      const result = await service.listNotebooks();
      return { content: [{ type: 'text', text: JSON.stringify(result, null, 2) }] };
    }
  );

  server.tool(
    'list_tags',
    'List all tags with guid and name',
    {},
    async () => {
      const result = await service.listTags();
      return { content: [{ type: 'text', text: JSON.stringify(result, null, 2) }] };
    }
  );
}
